src/cache.c

```diff
@@ -0,0 +1,201 @@
+//========================================================================
+// GLFW 3.4 - www.glfw.org
+//------------------------------------------------------------------------
+// Copyright (c) 2021 Emmanuel Gil Peyrot <linkmauve@linkmauve.fr>
+//
+// This software is provided 'as-is', without any express or implied
+// warranty. In no event will the authors be held liable for any damages
+// arising from the use of this software.
+//
+// Permission is granted to anyone to use this software for any purpose,
+// including commercial applications, and to alter it and redistribute it
+// freely, subject to the following restrictions:
+//
+// 1. The origin of this software must not be misrepresented; you must not
+//    claim that you wrote the original software. If you use this software
+//    in a product, an acknowledgment in the product documentation would
+//    be appreciated but is not required.
+//
+// 2. Altered source versions must be plainly marked as such, and must not
+//    be misrepresented as being the original software.
+//
+// 3. This notice may not be removed or altered from any source
+//    distribution.
+//
+//========================================================================
+// Please use C89 style variable declarations in this file because VS 2010
+//========================================================================
+
+#define _XOPEN_SOURCE 500
+
+#include "cache.h"
+
+#ifndef HAVE_XXHASH
+#error "libxxhash is required for the compose cache"
+#endif
+
+#include <fcntl.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/mman.h>
+#include <sys/stat.h>
+#include <unistd.h>
+#define XXH_INLINE_ALL
+#include <xxhash.h>
+
+static GLFWbool
+hashString(const char *string, size_t len, XXH128_canonical_t *out)
+{
+    XXH3_state_t *state;
+    XXH_errorcode err;
+    XXH128_hash_t hash;
+
+    state = XXH3_createState();
+    if (!state)
+        return GLFW_FALSE;
+
+    err = XXH3_128bits_reset(state);
+    if (err != XXH_OK)
+    {
+        XXH3_freeState(state);
+        return GLFW_FALSE;
+    }
+
+    err = XXH3_128bits_update(state, string, len);
+    if (err != XXH_OK)
+    {
+        XXH3_freeState(state);
+        return GLFW_FALSE;
+    }
+
+    hash = XXH3_128bits_digest(state);
+    XXH3_freeState(state);
+    XXH128_canonicalFromHash(out, hash);
+    return GLFW_TRUE;
+}
+
+GLFWbool
+cacheGetPathFromString(const char *string, char **cachePath)
+{
+    const char *home, *xdg;
+    char *path;
+    int length;
+    char *hash_part;
+    XXH128_canonical_t hash;
+
+    /* Mostly copied over from xkb_context_include_path_append_default(). */
+    home = getenv("HOME");
+    xdg = getenv("XDG_CACHE_HOME");
+    if (xdg != NULL)
+    {
+        length = strlen(xdg) + strlen("/glfw/XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX") + 1;
```

Comment:
Why is the filename `XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX` ?

Reply:
That is just a placeholder, I could have used `+ 32` instead (maybe that’s more readable?), the actual filename gets generated from the hash of the input on lines 118-122.

Reply:
I’ve just pushed a change with the `+ 32`, which is indeed less surprising.